File: src/main/ruby/g0001_0100/s0046_permutations/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Backtracking
# #Algorithm_I_Day_11_Recursion_Backtracking #Level_2_Day_20_Brute_Force/Backtracking
# #Udemy_Backtracking/Recursion #Big_O_Time_O(n*n!)_Space_O(n+n!)
# #2023_11_19_Time_62_ms_(89.60%)_Space_211.1_MB_(64.00%)

# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
  return [] if nums.nil? || nums.empty?

  final_result = []
  permute_recur(nums, final_result, [], Array.new(nums.length, false))
  final_result
end

private

def permute_recur(nums, final_result, curr_result, used)
  if curr_result.length == nums.length
    final_result << curr_result.dup
    return
  end

  nums.each_with_index do |num, i|
    next if used[i]

    curr_result << num
    used[i] = true
    permute_recur(nums, final_result, curr_result, used)
    used[i] = false
    curr_result.pop
  end
end

File: src/main/cpp/g0001_0100/s0046_permutations/Solution.cpp
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Backtracking
// #Algorithm_I_Day_11_Recursion_Backtracking #Level_2_Day_20_Brute_Force/Backtracking
// #Udemy_Backtracking/Recursion #Big_O_Time_O(n*n!)_Space_O(n+n!)
// #2024_05_25_Time_0_ms_(100.00%)_Space_9.6_MB_(50.28%)

#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> permute(std::vector<int>& nums) {
        if (nums.empty()) {
            return {};
        }
        std::vector<std::vector<int>> finalResult;
        std::vector<int> currResult;
        std::vector<bool> used(nums.size(), false);
        permuteRecur(nums, finalResult, currResult, used);
        return finalResult;
    }

private:
    void permuteRecur(const std::vector<int>& nums, 
                      std::vector<std::vector<int>>& finalResult, 
                      std::vector<int>& currResult, 
                      std::vector<bool>& used) {
        if (currResult.size() == nums.size()) {
            finalResult.push_back(currResult);
            return;
        }
        for (size_t i = 0; i < nums.size(); ++i) {
            if (used[i]) {
                continue;
            }
            currResult.push_back(nums[i]);
            used[i] = true;
            permuteRecur(nums, finalResult, currResult, used);
            used[i] = false;
            currResult.pop_back();
        }
    }
};
